lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -0,0 +1,1499 @@
+
+    Use this class if your charm's certificates are managed per unit.
+    """
+
+    on = CertificatesRequirerCharmEvents()  # type: ignore[reportAssignmentType]
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relationship_name: str,
+        certificate_requests: List[CertificateRequest],
+        mode: Mode = Mode.UNIT,
+        refresh_events: List[BoundEvent] = [],
+    ):
+        """Create a new instance of the TLSCertificatesRequiresV4 class.
+
+        Args:
+            charm (CharmBase): The charm instance to relate to.
+            relationship_name (str): The name of the relation that provides the certificates.
+            certificate_requests (List[CertificateRequest]): A list of certificate requests.
+            mode (Mode): The mode of the charm. Default is Mode.UNIT.
+            refresh_events (List[BoundEvent]): A list of events to trigger a refresh of
+              the certificates.
+        """
+        super().__init__(charm, relationship_name)
+        if not JujuVersion.from_environ().has_secrets:
+            logger.warning("This version of the TLS library requires Juju secrets (Juju >= 3.0)")
+        if not self._mode_is_valid(mode):
+            raise TLSCertificatesError("Invalid mode. Must be Mode.UNIT or Mode.APP")
+        self.charm = charm
+        self.relationship_name = relationship_name
+        self.certificate_requests = certificate_requests
+        self.mode = mode
+        self.framework.observe(self.on.update_status, self._configure)
+        self.framework.observe(charm.on[relationship_name].relation_created, self._configure)
+        self.framework.observe(charm.on[relationship_name].relation_changed, self._configure)
+        self.framework.observe(charm.on.secret_expired, self._on_secret_expired)
+        for event in refresh_events:
+            self.framework.observe(event, self._configure)
+
+    def _configure(self, _: EventBase):
+        """Handle TLS Certificates Relation Data.
+
+        This method is called during any TLS relation event.
+        It will generate a private key if it doesn't exist yet.
+        It will send certificate requests if they haven't been sent yet.
+        It will find available certificates and emit events.
+        """
+        if not self._tls_relation_created():
+            logger.debug("TLS relation not created yet.")
+            return
+        self._generate_private_key()
+        self._send_certificate_requests()
+        self._find_available_certificates()
+        self._cleanup_certificate_requests()
+
+    def _mode_is_valid(self, mode) -> bool:
+        return mode in [Mode.UNIT, Mode.APP]
+
+    def _on_secret_expired(self, event: SecretExpiredEvent) -> None:
+        """Handle Secret Expired Event.
+
+        Renews certificate requests and removes the expired secret.
+        """
+        if not event.secret.label or not event.secret.label.startswith(f"{LIBID}-certificate"):
```

Comment:
We might want to look into the expiry of the secret holding the private key too.

Reply:
This one never expires